### ISSUE TITLE
Nominal enum assignability after union

### ISSUE THREAD

**Issue by ivogabe** [Microsoft/TypeScript#10776]
**TypeScript Version:**  2.0 RC. This was working correctly in 2.0 beta.

**Code**

```ts
module X {
	export enum Foo {
		A, B
	}
}
module Y {
	export enum Foo {
		A, B
	}
}
const x: X.Foo = Y.Foo.A;
const y: X.Foo | undefined = Y.Foo.A;
```

**Expected behavior:**
Both assignments are valid.

**Actual behavior:** 
The first assignment is valid, the second throws an error saying "Type 'Foo' is not assignable to type 'Foo | undefined'."

This block a project that depends on the TypeScript api, and on some package that also depends on TypeScript. So TS is installed on `node_modules/typescript` and `node_modules/foo/node_modules/typescript`. (Normally NPM would install TypeScript only once, but `foo` is a symlink.) This causes that objects from one installation of TypeScript cannot be passed to the other. The following code illustrates that:
```typescript


import * as ts from "typescript";
import * as ts2 from "foo/node_modules/typescript";

const identifier: ts.Identifier = undefined as any;
const identifier2: ts2.Identifier = identifier;

const ok: ts.SyntaxKind = ts2.SyntaxKind.AbstractKeyword;
const error: ts.SyntaxKind | undefined = ts2.SyntaxKind.AbstractKeyword;
```
Error on the declaration of `identifier2`:
```
Type 'ts.Identifier' is not assignable to type 'ts.Identifier'.
  Types of property 'originalKeywordKind' are incompatible.
    Type 'SyntaxKind | undefined' is not assignable to type 'SyntaxKind | undefined'.
      Type 'SyntaxKind.Unknown' is not assignable to type 'SyntaxKind | undefined'.
```
And on `error`:
```
Type 'SyntaxKind' is not assignable to type 'SyntaxKind | undefined'.
```

**Comment by RyanCavanaugh** [comment]
@mhegazy milestone?

**Comment by sandersn** [comment]
Note: requires `"strictNullChecks": true` to repro

**Comment by sandersn** [comment]
repro without strictNullChecks: `const y: X.Foo | boolean = Y.Foo.A;`

**Comment by sandersn** [comment]
The fix is very similar to the one for #10738, so I'm going to fix both in #10784.
